### ISSUE TITLE
"Initialize Language Server" goes on forever

### ISSUE THREAD

**Issue by angelozerr** [angelozerr/lsp4e-freemarker#2]
On a new Photon M6 installation, if I open an `*.ftl` file, Eclipse shows "Initialize Language Server for test.ftl" in the lower right corner, with the progress bar animation, and that goes on forever. Thus there's no LSP4E related functionality in the editor.

In the error log I find this:

```
!ENTRY org.eclipse.lsp4e 4 0 2018-03-17 19:06:56.505
!MESSAGE LanguageServer not initialized after 10s
!STACK 0
java.util.concurrent.TimeoutException
	at java.util.concurrent.CompletableFuture.timedGet(Unknown Source)
	at java.util.concurrent.CompletableFuture.get(Unknown Source)
	at org.eclipse.lsp4e.LanguageServerWrapper.getServerCapabilities(LanguageServerWrapper.java:526)
	at org.eclipse.lsp4e.LanguageServiceAccessor.getLSWrappers(LanguageServiceAccessor.java:281)
	at org.eclipse.lsp4e.LanguageServiceAccessor.getLSPDocumentInfosFor(LanguageServiceAccessor.java:395)
	at org.eclipse.lsp4e.ConnectDocumentToLanguageServerSetupParticipant$1.run(ConnectDocumentToLanguageServerSetupParticipant.java:74)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:60)
```

So it's two problems:
- The language server initialization fails
- If the language server initialization fails, I should get some kind of error indicator, and the progress indicator should stop.

**Comment by angelozerr** [comment]
@ddekany please retry to install lsp4e-freemarker and tell me if this issue is fixed. Thanks!

**Comment by ddekany** [comment]
Still not working. Now there's no progress indicator like originally, but FTL errors aren't marked, and the whole(!) Eclipse goes unresponsive for about 30 seconds when opening the file. The error in the log says:

```
!ENTRY org.eclipse.lsp4e 4 0 2018-03-18 09:11:20.530
!MESSAGE java.util.concurrent.CancellationException
!STACK 0
java.util.concurrent.ExecutionException: java.util.concurrent.CancellationException
	at java.util.concurrent.CompletableFuture.reportGet(Unknown Source)
	at java.util.concurrent.CompletableFuture.get(Unknown Source)
	at org.eclipse.lsp4e.LanguageServerWrapper.getServerCapabilities(LanguageServerWrapper.java:526)
	at org.eclipse.lsp4e.LanguageServiceAccessor.getLSWrappers(LanguageServiceAccessor.java:281)
	at org.eclipse.lsp4e.LanguageServiceAccessor.getLSPDocumentInfosFor(LanguageServiceAccessor.java:395)
	at org.eclipse.lsp4e.ConnectDocumentToLanguageServerSetupParticipant$1.run(ConnectDocumentToLanguageServerSetupParticipant.java:74)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:60)
Caused by: java.util.concurrent.CancellationException
	at java.util.concurrent.CompletableFuture.cancel(Unknown Source)
	at org.eclipse.lsp4e.LanguageServerWrapper.stop(LanguageServerWrapper.java:292)
	at org.eclipse.lsp4e.LanguageServerWrapper.start(LanguageServerWrapper.java:169)
	at org.eclipse.lsp4e.LanguageServerWrapper.getInitializedServer(LanguageServerWrapper.java:495)
	at org.eclipse.lsp4e.LanguageServiceAccessor.lambda$2(LanguageServiceAccessor.java:140)
	at java.util.stream.ReferencePipeline$3$1.accept(Unknown Source)
	at java.util.Iterator.forEachRemaining(Unknown Source)
	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Unknown Source)
	at java.util.stream.AbstractPipeline.copyInto(Unknown Source)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source)
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(Unknown Source)
	at java.util.stream.AbstractPipeline.evaluate(Unknown Source)
	at java.util.stream.ReferencePipeline.collect(Unknown Source)
	at org.eclipse.lsp4e.LanguageServiceAccessor.getInitializedLanguageServers(LanguageServiceAccessor.java:147)
	at org.eclipse.lsp4e.HasLanguageServerPropertyTester.test(HasLanguageServerPropertyTester.java:32)
	at org.eclipse.core.internal.expressions.Property.test(Property.java:58)
	at org.eclipse.core.internal.expressions.TestExpression.evaluate(TestExpression.java:100)
	at org.eclipse.core.internal.expressions.CompositeExpression.evaluateAnd(CompositeExpression.java:51)
	at org.eclipse.core.internal.expressions.WithExpression.evaluate(WithExpression.java:75)
	at org.eclipse.ui.internal.texteditor.codemining.CodeMiningProviderDescriptor.matches(CodeMiningProviderDescriptor.java:185)
	at org.eclipse.ui.internal.texteditor.codemining.CodeMiningProviderRegistry.getProviders(CodeMiningProviderRegistry.java:108)
	at org.eclipse.ui.texteditor.AbstractTextEditor.installCodeMinigProviders(AbstractTextEditor.java:4046)
	at org.eclipse.ui.texteditor.AbstractTextEditor.initializeSourceViewer(AbstractTextEditor.java:4036)
	at org.eclipse.ui.texteditor.AbstractTextEditor.createPartControl(AbstractTextEditor.java:3475)
	at org.eclipse.ui.texteditor.StatusTextEditor.createPartControl(StatusTextEditor.java:61)
	at org.eclipse.ui.texteditor.AbstractDecoratedTextEditor.createPartControl(AbstractDecoratedTextEditor.java:444)
	at org.eclipse.ui.internal.genericeditor.ExtensionBasedTextEditor.createPartControl(ExtensionBasedTextEditor.java:67)
	at org.eclipse.ui.internal.e4.compatibility.CompatibilityPart.createPartControl(CompatibilityPart.java:151)
	at org.eclipse.ui.internal.e4.compatibility.CompatibilityEditor.createPartControl(CompatibilityEditor.java:99)
	at org.eclipse.ui.internal.e4.compatibility.CompatibilityPart.create(CompatibilityPart.java:355)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at org.eclipse.e4.core.internal.di.MethodRequestor.execute(MethodRequestor.java:55)
	at org.eclipse.e4.core.internal.di.InjectorImpl.processAnnotated(InjectorImpl.java:1002)
	at org.eclipse.e4.core.internal.di.InjectorImpl.processAnnotated(InjectorImpl.java:967)
	at org.eclipse.e4.core.internal.di.InjectorImpl.internalInject(InjectorImpl.java:136)
	at org.eclipse.e4.core.internal.di.InjectorImpl.internalMake(InjectorImpl.java:411)
	at org.eclipse.e4.core.internal.di.InjectorImpl.make(InjectorImpl.java:330)
	at org.eclipse.e4.core.contexts.ContextInjectionFactory.make(ContextInjectionFactory.java:190)
	at org.eclipse.e4.ui.internal.workbench.ReflectionContributionFactory.createFromBundle(ReflectionContributionFactory.java:105)
	at org.eclipse.e4.ui.internal.workbench.ReflectionContributionFactory.doCreate(ReflectionContributionFactory.java:74)
	at org.eclipse.e4.ui.internal.workbench.ReflectionContributionFactory.create(ReflectionContributionFactory.java:56)
	at org.eclipse.e4.ui.workbench.renderers.swt.ContributedPartRenderer.createWidget(ContributedPartRenderer.java:129)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.createWidget(PartRenderingEngine.java:992)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.safeCreateGui(PartRenderingEngine.java:661)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.safeCreateGui(PartRenderingEngine.java:767)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.access$0(PartRenderingEngine.java:738)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$2.run(PartRenderingEngine.java:732)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:42)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.createGui(PartRenderingEngine.java:716)
	at org.eclipse.e4.ui.internal.workbench.PartServiceImpl$1.handleEvent(PartServiceImpl.java:104)
	at org.eclipse.e4.ui.services.internal.events.UIEventHandler$1.run(UIEventHandler.java:40)
	at org.eclipse.swt.widgets.Synchronizer.syncExec(Synchronizer.java:233)
	at org.eclipse.ui.internal.UISynchronizer.syncExec(UISynchronizer.java:144)
	at org.eclipse.swt.widgets.Display.syncExec(Display.java:4894)
	at org.eclipse.e4.ui.internal.workbench.swt.E4Application$1.syncExec(E4Application.java:212)
	at org.eclipse.e4.ui.services.internal.events.UIEventHandler.handleEvent(UIEventHandler.java:36)
	at org.eclipse.equinox.internal.event.EventHandlerWrapper.handleEvent(EventHandlerWrapper.java:201)
	at org.eclipse.equinox.internal.event.EventHandlerTracker.dispatchEvent(EventHandlerTracker.java:196)
	at org.eclipse.equinox.internal.event.EventHandlerTracker.dispatchEvent(EventHandlerTracker.java:1)
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:230)
	at org.eclipse.osgi.framework.eventmgr.ListenerQueue.dispatchEventSynchronous(ListenerQueue.java:148)
	at org.eclipse.equinox.internal.event.EventAdminImpl.dispatchEvent(EventAdminImpl.java:135)
	at org.eclipse.equinox.internal.event.EventAdminImpl.sendEvent(EventAdminImpl.java:78)
	at org.eclipse.equinox.internal.event.EventComponent.sendEvent(EventComponent.java:39)
	at org.eclipse.e4.ui.services.internal.events.EventBroker.send(EventBroker.java:52)
	at org.eclipse.e4.ui.internal.workbench.UIEventPublisher.notifyChanged(UIEventPublisher.java:60)
	at org.eclipse.emf.common.notify.impl.BasicNotifierImpl.eNotify(BasicNotifierImpl.java:374)
	at org.eclipse.e4.ui.model.application.ui.impl.ElementContainerImpl.setSelectedElement(ElementContainerImpl.java:174)
	at org.eclipse.e4.ui.internal.workbench.ModelServiceImpl.showElementInWindow(ModelServiceImpl.java:624)
	at org.eclipse.e4.ui.internal.workbench.ModelServiceImpl.bringToTop(ModelServiceImpl.java:588)
	at org.eclipse.e4.ui.internal.workbench.PartServiceImpl.delegateBringToTop(PartServiceImpl.java:769)
	at org.eclipse.e4.ui.internal.workbench.PartServiceImpl.bringToTop(PartServiceImpl.java:401)
	at org.eclipse.e4.ui.internal.workbench.PartServiceImpl.showPart(PartServiceImpl.java:1219)
	at org.eclipse.ui.internal.WorkbenchPage.busyOpenEditor(WorkbenchPage.java:3277)
	at org.eclipse.ui.internal.WorkbenchPage.access$26(WorkbenchPage.java:3192)
	at org.eclipse.ui.internal.WorkbenchPage$10.run(WorkbenchPage.java:3174)
	at org.eclipse.swt.custom.BusyIndicator.showWhile(BusyIndicator.java:70)
	at org.eclipse.ui.internal.WorkbenchPage.openEditor(WorkbenchPage.java:3169)
	at org.eclipse.ui.internal.WorkbenchPage.openEditor(WorkbenchPage.java:3133)
	at org.eclipse.ui.internal.WorkbenchPage.openEditor(WorkbenchPage.java:3123)
	at org.eclipse.ui.ide.IDE.openEditor(IDE.java:563)
	at org.eclipse.ui.ide.IDE.openEditor(IDE.java:518)
	at org.eclipse.jdt.internal.ui.javaeditor.EditorUtility.openInEditor(EditorUtility.java:360)
	at org.eclipse.jdt.internal.ui.javaeditor.EditorUtility.openInEditor(EditorUtility.java:167)
	at org.eclipse.jdt.ui.actions.OpenAction.run(OpenAction.java:284)
	at org.eclipse.jdt.ui.actions.OpenAction.run(OpenAction.java:249)
	at org.eclipse.jdt.ui.actions.SelectionDispatchAction.dispatchRun(SelectionDispatchAction.java:271)
	at org.eclipse.jdt.ui.actions.SelectionDispatchAction.run(SelectionDispatchAction.java:249)
	at org.eclipse.jdt.internal.ui.navigator.OpenAndExpand.run(OpenAndExpand.java:50)
	at org.eclipse.ui.actions.RetargetAction.run(RetargetAction.java:216)
	at org.eclipse.ui.navigator.CommonNavigatorManager$2.open(CommonNavigatorManager.java:191)
	at org.eclipse.ui.OpenAndLinkWithEditorHelper$InternalListener.open(OpenAndLinkWithEditorHelper.java:46)
	at org.eclipse.jface.viewers.StructuredViewer$2.run(StructuredViewer.java:850)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:42)
	at org.eclipse.ui.internal.JFaceUtil.lambda$0(JFaceUtil.java:44)
	at org.eclipse.jface.util.SafeRunnable.run(SafeRunnable.java:173)
	at org.eclipse.jface.viewers.StructuredViewer.fireOpen(StructuredViewer.java:847)
	at org.eclipse.jface.viewers.StructuredViewer.handleOpen(StructuredViewer.java:1162)
	at org.eclipse.ui.navigator.CommonViewer.handleOpen(CommonViewer.java:451)
	at org.eclipse.jface.util.OpenStrategy.fireOpenEvent(OpenStrategy.java:273)
	at org.eclipse.jface.util.OpenStrategy.access$2(OpenStrategy.java:268)
	at org.eclipse.jface.util.OpenStrategy$1.handleEvent(OpenStrategy.java:308)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:86)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4433)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1086)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4243)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3822)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1150)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:336)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1039)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:153)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:681)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:336)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:595)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:148)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:152)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:196)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:134)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:388)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:243)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:656)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:592)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1498)

!ENTRY org.eclipse.lsp4e 4 0 2018-03-18 09:11:31.610
!MESSAGE LanguageServer not initialized after 10s
!STACK 0
java.util.concurrent.TimeoutException
	at java.util.concurrent.CompletableFuture.timedGet(Unknown Source)
	at java.util.concurrent.CompletableFuture.get(Unknown Source)
	at org.eclipse.lsp4e.LanguageServerWrapper.getServerCapabilities(LanguageServerWrapper.java:526)
	at org.eclipse.lsp4e.LanguageServiceAccessor.lambda$8(LanguageServiceAccessor.java:352)
	at java.util.stream.ReferencePipeline$2$1.accept(Unknown Source)
	at java.util.stream.ReferencePipeline$2$1.accept(Unknown Source)
	at java.util.HashMap$KeySpliterator.forEachRemaining(Unknown Source)
	at java.util.stream.AbstractPipeline.copyInto(Unknown Source)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source)
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(Unknown Source)
	at java.util.stream.AbstractPipeline.evaluate(Unknown Source)
	at java.util.stream.ReferencePipeline.collect(Unknown Source)
	at org.eclipse.lsp4e.LanguageServiceAccessor.getMatchingStartedWrappers(LanguageServiceAccessor.java:354)
	at org.eclipse.lsp4e.LanguageServiceAccessor.getLSWrappers(LanguageServiceAccessor.java:264)
	at org.eclipse.lsp4e.LanguageServiceAccessor.getLSPDocumentInfosFor(LanguageServiceAccessor.java:395)
	at org.eclipse.lsp4e.outline.EditorToOutlineAdapterFactory.getAdapter(EditorToOutlineAdapterFactory.java:34)
	at org.eclipse.core.internal.adapter.AdapterFactoryProxy.getAdapter(AdapterFactoryProxy.java:83)
	at org.eclipse.core.internal.runtime.AdapterManager.getAdapter(AdapterManager.java:295)
	at org.eclipse.ui.part.WorkbenchPart.getAdapter(WorkbenchPart.java:142)
	at org.eclipse.ui.texteditor.AbstractTextEditor.getAdapter(AbstractTextEditor.java:6213)
	at org.eclipse.ui.texteditor.AbstractDecoratedTextEditor.getAdapter(AbstractDecoratedTextEditor.java:1417)
	at org.eclipse.ui.editors.text.TextEditor.getAdapter(TextEditor.java:160)
	at org.eclipse.core.runtime.Adapters.adapt(Adapters.java:59)
	at org.eclipse.core.runtime.Adapters.adapt(Adapters.java:107)
	at org.eclipse.ui.views.contentoutline.ContentOutline.doCreatePage(ContentOutline.java:129)
	at org.eclipse.ui.part.PageBookView.createPage(PageBookView.java:375)
	at org.eclipse.ui.part.PageBookView.partActivated(PageBookView.java:715)
	at org.eclipse.ui.views.contentoutline.ContentOutline.partBroughtToTop(ContentOutline.java:204)
	at org.eclipse.ui.part.PageBookView$1.partBroughtToTop(PageBookView.java:1033)
	at org.eclipse.ui.internal.WorkbenchPage$27.run(WorkbenchPage.java:5293)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:42)
	at org.eclipse.ui.internal.WorkbenchPage.firePartBroughtToTop(WorkbenchPage.java:5290)
	at org.eclipse.ui.internal.WorkbenchPage.access$14(WorkbenchPage.java:5271)
	at org.eclipse.ui.internal.WorkbenchPage$E4PartListener.partBroughtToTop(WorkbenchPage.java:221)
	at org.eclipse.e4.ui.internal.workbench.PartServiceImpl$7.run(PartServiceImpl.java:314)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:42)
	at org.eclipse.e4.ui.internal.workbench.PartServiceImpl.firePartBroughtToTop(PartServiceImpl.java:311)
	at org.eclipse.e4.ui.internal.workbench.PartServiceImpl.access$4(PartServiceImpl.java:309)
	at org.eclipse.e4.ui.internal.workbench.PartServiceImpl$1.handleEvent(PartServiceImpl.java:106)
	at org.eclipse.e4.ui.services.internal.events.UIEventHandler$1.run(UIEventHandler.java:40)
	at org.eclipse.swt.widgets.Synchronizer.syncExec(Synchronizer.java:233)
	at org.eclipse.ui.internal.UISynchronizer.syncExec(UISynchronizer.java:144)
	at org.eclipse.swt.widgets.Display.syncExec(Display.java:4894)
	at org.eclipse.e4.ui.internal.workbench.swt.E4Application$1.syncExec(E4Application.java:212)
	at org.eclipse.e4.ui.services.internal.events.UIEventHandler.handleEvent(UIEventHandler.java:36)
	at org.eclipse.equinox.internal.event.EventHandlerWrapper.handleEvent(EventHandlerWrapper.java:201)
	at org.eclipse.equinox.internal.event.EventHandlerTracker.dispatchEvent(EventHandlerTracker.java:196)
	at org.eclipse.equinox.internal.event.EventHandlerTracker.dispatchEvent(EventHandlerTracker.java:1)
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:230)
	at org.eclipse.osgi.framework.eventmgr.ListenerQueue.dispatchEventSynchronous(ListenerQueue.java:148)
	at org.eclipse.equinox.internal.event.EventAdminImpl.dispatchEvent(EventAdminImpl.java:135)
	at org.eclipse.equinox.internal.event.EventAdminImpl.sendEvent(EventAdminImpl.java:78)
	at org.eclipse.equinox.internal.event.EventComponent.sendEvent(EventComponent.java:39)
	at org.eclipse.e4.ui.services.internal.events.EventBroker.send(EventBroker.java:52)
	at org.eclipse.e4.ui.internal.workbench.UIEventPublisher.notifyChanged(UIEventPublisher.java:60)
	at org.eclipse.emf.common.notify.impl.BasicNotifierImpl.eNotify(BasicNotifierImpl.java:374)
	at org.eclipse.e4.ui.model.application.ui.impl.ElementContainerImpl.setSelectedElement(ElementContainerImpl.java:174)
	at org.eclipse.e4.ui.internal.workbench.ModelServiceImpl.showElementInWindow(ModelServiceImpl.java:624)
	at org.eclipse.e4.ui.internal.workbench.ModelServiceImpl.bringToTop(ModelServiceImpl.java:588)
	at org.eclipse.e4.ui.internal.workbench.PartServiceImpl.delegateBringToTop(PartServiceImpl.java:769)
	at org.eclipse.e4.ui.internal.workbench.PartServiceImpl.bringToTop(PartServiceImpl.java:401)
	at org.eclipse.e4.ui.internal.workbench.PartServiceImpl.showPart(PartServiceImpl.java:1219)
	at org.eclipse.ui.internal.WorkbenchPage.busyOpenEditor(WorkbenchPage.java:3277)
	at org.eclipse.ui.internal.WorkbenchPage.access$26(WorkbenchPage.java:3192)
	at org.eclipse.ui.internal.WorkbenchPage$10.run(WorkbenchPage.java:3174)
	at org.eclipse.swt.custom.BusyIndicator.showWhile(BusyIndicator.java:70)
	at org.eclipse.ui.internal.WorkbenchPage.openEditor(WorkbenchPage.java:3169)
	at org.eclipse.ui.internal.WorkbenchPage.openEditor(WorkbenchPage.java:3133)
	at org.eclipse.ui.internal.WorkbenchPage.openEditor(WorkbenchPage.java:3123)
	at org.eclipse.ui.ide.IDE.openEditor(IDE.java:563)
	at org.eclipse.ui.ide.IDE.openEditor(IDE.java:518)
	at org.eclipse.jdt.internal.ui.javaeditor.EditorUtility.openInEditor(EditorUtility.java:360)
	at org.eclipse.jdt.internal.ui.javaeditor.EditorUtility.openInEditor(EditorUtility.java:167)
	at org.eclipse.jdt.ui.actions.OpenAction.run(OpenAction.java:284)
	at org.eclipse.jdt.ui.actions.OpenAction.run(OpenAction.java:249)
	at org.eclipse.jdt.ui.actions.SelectionDispatchAction.dispatchRun(SelectionDispatchAction.java:271)
	at org.eclipse.jdt.ui.actions.SelectionDispatchAction.run(SelectionDispatchAction.java:249)
	at org.eclipse.jdt.internal.ui.navigator.OpenAndExpand.run(OpenAndExpand.java:50)
	at org.eclipse.ui.actions.RetargetAction.run(RetargetAction.java:216)
	at org.eclipse.ui.navigator.CommonNavigatorManager$2.open(CommonNavigatorManager.java:191)
	at org.eclipse.ui.OpenAndLinkWithEditorHelper$InternalListener.open(OpenAndLinkWithEditorHelper.java:46)
	at org.eclipse.jface.viewers.StructuredViewer$2.run(StructuredViewer.java:850)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:42)
	at org.eclipse.ui.internal.JFaceUtil.lambda$0(JFaceUtil.java:44)
	at org.eclipse.jface.util.SafeRunnable.run(SafeRunnable.java:173)
	at org.eclipse.jface.viewers.StructuredViewer.fireOpen(StructuredViewer.java:847)
	at org.eclipse.jface.viewers.StructuredViewer.handleOpen(StructuredViewer.java:1162)
	at org.eclipse.ui.navigator.CommonViewer.handleOpen(CommonViewer.java:451)
	at org.eclipse.jface.util.OpenStrategy.fireOpenEvent(OpenStrategy.java:273)
	at org.eclipse.jface.util.OpenStrategy.access$2(OpenStrategy.java:268)
	at org.eclipse.jface.util.OpenStrategy$1.handleEvent(OpenStrategy.java:308)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:86)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4433)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1086)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4243)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3822)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1150)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:336)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1039)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:153)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:681)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:336)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:595)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:148)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:152)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:196)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:134)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:388)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:243)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:656)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:592)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1498)
```

Such errors occur repeatedly, around 10 s between.

**Comment by angelozerr** [comment]
@ddekany please retry to install and I hope really it will works now.

The change that I have done is to unzip the plugin in the Eclipse_HOME/plugins

So you should have now this folder after installation` ECLIPSE_HOME\plugins\org.eclipse.lsp4e.freemarker_1.0.0.201803190625` which contains the JAR `server/freemarker-languageserver-all.jar` which is the Freemarker Lanaguage Server.

If it doesn't work:

 * have you spaces in your ECLIPSE_HOME?
 * please try with Lanuch Configuration (explained after)

1) create a Jav aproject my-freemarker_languageserver
2) Add ECLIPSE_HOME\plugins\org.eclipse.lsp4e.freemarker_1.0.0.201803190625/server/freemarker-languageserver-all.jar`  in the class path:

![image](https://user-images.githubusercontent.com/1932211/37581472-1d63a512-2b49-11e8-8532-42636771798c.png)

3) Run it to generate a launch (FreemarkerServerStdioLauncher)

![image](https://user-images.githubusercontent.com/1932211/37581524-447d0468-2b49-11e8-9a8b-f5908b683d0a.png)

4) Stop the launch
5) Bind the launch with Freemarker content type

With Language Server preferences:

![image](https://user-images.githubusercontent.com/1932211/37581607-9458268e-2b49-11e8-9c8d-4518e52f333f.png)

You should have that:

![image](https://user-images.githubusercontent.com/1932211/37581666-d0dc6804-2b49-11e8-8331-e959d507c9df.png)

(The default Freemarker language server on the top must be desactivated)

6) Try

You should see that with trace console of LSP:

![image](https://user-images.githubusercontent.com/1932211/37581705-02605084-2b4a-11e8-8b4d-5d2e6b69859a.png)

**Comment by ddekany** [comment]
Still the same problem... The jar is there in the plugin directory, and I have no space or any special character in the eclipse home directory. (I have no `ECLIPSE_HOME` environment variable set manually by me, but I assume it's not what you meant.)

As of the launch configuration trick, when I try to select the launch configuration for the Launch Server (as on your screen shot), I can't, as the "OK" button is grayed out (and yes, I did select "run" from the dropdown).

Anyway, it's quite certain that the JVM is not launched, as I don't see any extra `java.exe`-s (except if I launch the jar manually, of course). So, even regardless of this issue, it would be important that it's reported somewhere why the language server startup was failed. Can't you log it somewhere? Because surely I'm not the last one who runs into some obscure errors. (Also, it's very strange that a language server can't just be started within the existing VM of Eclipse... well, I assume that's a limitation of LSP4E.) 

Also that a failure to start the Language Server causes Eclipse to freeze for a minute or so is quite strange. I mean a Language Server does its job asynchronously, so things should just naturally go on without it.

**Comment by angelozerr** [comment]
> Still the same problem... 

-(

> The jar is there in the plugin directory, and I have no space or any special character in the eclipse home directory. 

Ok cool.

> (I have no ECLIPSE_HOME environment variable set manually by me, but I assume it's not what you meant.)

I meant YOUR_ECLIPSE_HOME, but it works you have the jar.

> As of the launch configuration trick, when I try to select the launch configuration for the Launch Server (as on your screen shot), I can't, as the "OK" button is grayed out (and yes, I did select "run" from the dropdown).

Have you selected the Freemarker content type (on the left combo) too like 

![](https://user-images.githubusercontent.com/1932211/37581607-9458268e-2b49-11e8-9c8d-4518e52f333f.png)

> So, even regardless of this issue, it would be important that it's reported somewhere why the language server startup was failed. Can't you log it somewhere? Because surely I'm not the last one who runs into some obscure errors. (Also, it's very strange that a language server can't just be started within the existing VM of Eclipse... well, I assume that's a limitation of LSP4E.)

To be honnest with you, I don't know how to do that, perhaps it's an issue for LSP4E. The code is very basic https://github.com/angelozerr/lsp4e-freemarker/blob/master/org.eclipse.lsp4e.freemarker/src/org/eclipse/lsp4e/freemarker/FreemarkerLanguageServer.java#L43

It executes the command `java -jar YOUR_ECLIPSE_HOME/......../server/freemarker-languageserver-all.jar`

If you open your OS Shell terminal and you do that, is it working?

> Also that a failure to start the Language Server causes Eclipse to freeze for a minute or so is quite strange. I mean a Language Server does its job asynchronously, so things should just naturally go on without it.

It's an issue for LSP4e.

**Comment by ddekany** [comment]
Ugh... yeah, seems I haven't selected a Content Type. So now I could press "OK", but it still does the same Eclipse freezing thing. If I delete my Launch Configuration from under Language Servers, then it immediately opens the *.ftll files, so it's surely freezes because of the attempted launching of the language server. What's very odd that it works for you... do you have some other version of SLF4E of something?

And yes, I can start the server from the command line.

Anyway, at this point, I guess I will have to debug this. Not yet sure how it's done, but I remember somehow I did debug the JBoss FreeMarker IDE back then... maybe it's the same with this. I will try to do that in the coming days, hopefully.

**Comment by angelozerr** [comment]
> so it's surely freezes because of the attempted launching of the language server

Indeed your trace comes from LSP4E.

> What's very odd that it works for you... do you have some other version of SLF4E of something?

No I'm with Windows 64 bits I  have tried it to download Photon M6 and install lsp4e-freemarker with the update site and it works well.

> And yes, I can start the server from the command line.

It's so crazy!

> Anyway, at this point, I guess I will have to debug this. 

Yes sure. Now you have installed dependencies, it will be more easy to debug the plugin:

 * clone lsp4e-freemarker
 * import only the project org.eclipse.lsp4e.freemarker in your workspace:

![image](https://user-images.githubusercontent.com/1932211/37625134-fd9a8016-2bca-11e8-8069-82f41380ceb6.png)

 * open the MANIFEST.MF and click on the `Debug icon` on the top right of the editor:

![image](https://user-images.githubusercontent.com/1932211/37625187-39a2710e-2bcb-11e8-92a1-c9b070fba1db.png)


 * it will start a second eclipse instance which load the plugins. You can set a breakpint in the FreemarkerLanguageServer and open an ftl file in the second eclipse. You will have that:

![image](https://user-images.githubusercontent.com/1932211/37625348-d6b9dc3e-2bcb-11e8-9713-038b430957a8.png)

**Comment by angelozerr** [comment]
@ddekany I think you should try too to debug the Freemarker Language Server. See https://github.com/angelozerr/lsp4e-freemarker/wiki/Debug

**Comment by angelozerr** [comment]
@ddekany once https://opensagres.ci.cloudbees.com/job/lsp4e-freemarker/18/ will be finished, please reinstall it.

Now you should not have freeze (I hope) and you should see in the Error log the error why Freemarker Language Server cannot be

**Comment by ddekany** [comment]
We were working in parallel... I just wanted to say that the `freemarker-languageserver` debugging didn't help, because the problem is that it doesn't start at all, because (and I quote what your new code has logged - that one sub-issue fixed so far, as it gets logged):

```
!ENTRY org.eclipse.lsp4e 4 0 2018-03-21 20:16:17.211
!MESSAGE Unable to start language server: FreeMarker (org.eclipse.lsp4e.freemarker.FreemarkerLanguageServer@ecf6a1d9)
!STACK 0
java.io.IOException: Unable to start language server: FreeMarker (org.eclipse.lsp4e.freemarker.FreemarkerLanguageServer@ecf6a1d9)
	at org.eclipse.lsp4e.freemarker.FreemarkerLanguageServer.start(FreemarkerLanguageServer.java:141)
	at org.eclipse.lsp4e.LanguageServerWrapper.start(LanguageServerWrapper.java:174)
	at org.eclipse.lsp4e.LanguageServerWrapper.getInitializedServer(LanguageServerWrapper.java:495)
	at org.eclipse.lsp4e.LanguageServerWrapper.getServer(LanguageServerWrapper.java:479)
	at org.eclipse.lsp4e.LanguageServiceAccessor.getLSPDocumentInfosFor(LanguageServiceAccessor.java:398)
	at org.eclipse.lsp4e.operations.codelens.CodeLensProvider.provideCodeMinings(CodeLensProvider.java:33)
	at org.eclipse.lsp4e.operations.codelens.CodeLensProvider.provideCodeMinings(CodeLensProvider.java:63)
	at org.eclipse.jface.internal.text.codemining.CodeMiningManager.lambda$1(CodeMiningManager.java:154)
	at java.util.stream.ReferencePipeline$3$1.accept(Unknown Source)
	at java.util.Spliterators$ArraySpliterator.forEachRemaining(Unknown Source)
	at java.util.stream.AbstractPipeline.copyInto(Unknown Source)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source)
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(Unknown Source)
	at java.util.stream.AbstractPipeline.evaluate(Unknown Source)
	at java.util.stream.ReferencePipeline.collect(Unknown Source)
	at org.eclipse.jface.internal.text.codemining.CodeMiningManager.getCodeMinings(CodeMiningManager.java:155)
	at org.eclipse.jface.internal.text.codemining.CodeMiningManager.updateCodeMinings(CodeMiningManager.java:121)
	at org.eclipse.jface.internal.text.codemining.CodeMiningManager.run(CodeMiningManager.java:110)
	at org.eclipse.jface.text.source.SourceViewer.updateCodeMinings(SourceViewer.java:1310)
	at org.eclipse.jface.text.codemining.CodeMiningStrategy.initialReconcile(CodeMiningStrategy.java:44)
	at org.eclipse.jface.text.reconciler.Reconciler.initialProcess(Reconciler.java:189)
	at org.eclipse.jface.text.reconciler.AbstractReconciler$BackgroundThread.run(AbstractReconciler.java:174)
Caused by: java.io.IOException: Error: Registry key 'Software\JavaSoft\Java Runtime Environment'\CurrentVersion'
	at org.eclipse.lsp4e.freemarker.FreemarkerLanguageServer$GetErrorThread.check(FreemarkerLanguageServer.java:104)
	at org.eclipse.lsp4e.freemarker.FreemarkerLanguageServer.start(FreemarkerLanguageServer.java:138)
	... 21 more
```

The good news is that it doesn't freeze Eclipse this way (one more sub-issue fixed). (The not so good news that the average users still sees nothing about why there's no error highlighting - but, everything on its time.)

Fun fact: To find out what's going on, I have set up an Eclipse for Committers and started debugging `lps4e-freemarker`, and... there it works from some reason. Error markers are working. No errors, nothing to debug.

Before diving into this Java startup issue even deeper, do we really have to launch a new process? Even if we manage to figure out my issue, this always will be a very error prone thing. We are already inside a JVM, and as far as I see all that's needed is two streams to communicate with the language server. The other end of those streams could be "local" as well.

**Comment by angelozerr** [comment]
> I just wanted to say that the freemarker-languageserver debugging didn't help, because the problem is that it doesn't start at all, because (and I quote what your new code has logged - that one sub-issue fixed so far, as it gets logged):

Indeed, I have fixed the problem with start server with Process and not with Launch so you have the freeze problem.

> The good news is that it doesn't freeze Eclipse this way (one more sub-issue fixed). (The not so good news that the average users still sees nothing about why there's no error highlighting - but, everything on its time.)

Yes my fix is to wait 200ms and track the error stream of the Java Process to throw an IOException if Process error stream have an error. In your case you have:

```
Caused by: java.io.IOException: Error: Registry key 'Software\JavaSoft\Java Runtime Environment'\CurrentVersion'
	at org.eclipse.lsp4e.freemarker.FreemarkerLanguageServer$GetErrorThread.check(FreemarkerLanguageServer.java:104)
```

Now we can see your error. My code is very simple, it execute un the Java Process the command 

`java -jar ...freemarker-language-server-all.jar`

If I understand your error, it means that Eclipse starts with a some JDK/JRE Java8 and if you open a Shell terminal and you type:

`java --version`

You have not the same version, is that?

**Comment by ddekany** [comment]
Right. With `java` the shell finds Java 9 (it's what in the path). The registry key points to Java 8 (I don't know why, as Java 9 was installed last, with the official installer), and Java 9 doesn't even have an entry under "Java Runtime Environment" in the registry. Eclipse runs with Java 8 because (apparently) it relies on the `JAVA_HOME` environment variable, which points to Java 8.

However, even if I (a) Start Eclipse with Java 9 or (b) ensure that `java` runs Java 8, the same problem persists. But if you are sure that my problem is related to having different Java versions around, I can continue with uninstalling Java 9, just to be sure it's not found on some strange way.

Why can using two different Java versions can cause problem here? The Language Server is a separate process. It might as well a non-Java application.

Also, I would like to repeat that the best would be if we can just run the language server inside the same JVM as Eclipse. Launching `java` was never a fool proof thing... and I bet a lot of Language Servers are written in Java, so it's curious if LSP4E doesn't support that.

**Comment by angelozerr** [comment]
> However, even if I (a) Start Eclipse with Java 9 or (b) ensure that java runs Java 8, the same problem persists. But if you are sure that my problem is related to having different Java versions around, I can continue with uninstalling Java 9, just to be sure it's not found on some strange way.

No please keep your configuration. It's a very good usecase.

> Why can using two different Java versions can cause problem here? The Language Server is a separate process. It might as well a non-Java application.

It's because, I started the Freemarker Language Serevr Main with "java" command. Please reinstall lsp4e-freemarker, now I'm searching the java from the "Installed JRE" marked as default. 

![image](https://user-images.githubusercontent.com/1932211/37762770-8e4ecec0-2dbd-11e8-88a6-0747c28fc1a7.png)

I don't know (for the moment) how to use the JRE which starts Eclipse.

Please note you must restart your Eclipse if you change the "Installed JRE". 

Hope it will work!

**Comment by ddekany** [comment]
Nope, still the same problem... But it's still an improvement that it now gets the `java` path from there.

> It's because, I started the Freemarker Language Serevr Main with "java" command. 

Yes, that was clear. What's not clear is if why would it care if it uses a different java version. It's a separate OS process. How can this difference matter at all?

**Comment by angelozerr** [comment]
> Nope, still the same problem... But it's still an improvement that it now gets the java path from there.

Your Installed JRE is java8?

**Comment by ddekany** [comment]
Yes, Java 8 build 144, 64 bit.

**Comment by angelozerr** [comment]
And what is your error log?

**Comment by ddekany** [comment]
Sorry, I was hasty again... your commit wasn't applied. Actually it works with `getDefaultVMInstall` and all that.

May I ask why? I'm just curious... one `java` should be as good as the other.

**Comment by ddekany** [comment]
Also I have second thoughts regarding the correctness of this solution. Let's say we require Java 8 for the language server (which is quite probable). What if someone has an earlier Java set as the default in the Eclipse properties? It's also quite surprising fur the user that that setting influences if the language server can start. After all, the default is about what Java version you develop against, not what Java version the plugins use.

**Comment by angelozerr** [comment]
> Sorry, I was hasty again... actually it works with getDefaultVMInstall and all that.

It's a very cool news!

> May I ask why? I'm just curious... one java should be as good as the other.

To be honnest with you, I don't know why it doesn't work with java9. If you start the FreemarkerServerStdioLauncher with Run As and java9, is it working?

I close this issue since now you can start the Freemarker Language Server. Now the interesting work can start and implement other command like outline, higlight, etc and it will work with VSCode too!

I'm working on https://github.com/angelozerr/freemarker-languageserver/issues/1: I'm trying to hack FMParser to have a tolerant parser. Hope we will work together about this hard issue.

I close this issue since it works for you.

**Comment by ddekany** [comment]
This problem is not related Java 9. Just to be 100% sure, I have now uninstalled Java 9, even checked that it's actually deleted. Yet, the registry problem still pops up if the plugin just uses `java` instead of `getDefaultVMInstall` etc.

`freemarker-languageserver` can be started with Java 9 from the command-line. But from under Eclipse with a launch configuration it crashes with "Error occurred during initialization of boot layer
java.lang.LayerInstantiationException: Package jdk.internal.jimage.decompressor in both module jrt.fs and module java.base". I didn't try to fix that, as it looks like a separate issue that's not relevant for us (so far...).

Anyway, OK, let's close this and do something more productive for now. But it's not over because of the issues I have noted above: the default Java installation might be too low version, and why do we need to launch a new process anyway... well, later.
